File: 2024_d3/a.py
import re
# write regex to extract mul(number,number)

def part1():
    regex = re.compile(r'mul\((\d+),(\d+)\)')
    
    sum = 0
    with open ("input.txt", 'r') as f:
        for line in f:
            for match in regex.finditer(line):
                sum += int(match.group(1)) * int(match.group(2))
    
    print(sum)

part1()
def part2_line(line):

    regex = re.compile(r'mul\((\d+),(\d+)\)|don\'t\(\)|do\(\)')
    sum = 0
    enable = True
    for match in regex.finditer(line):
        print(match.group(0))
        if match.group(0).startswith('mul'):
            if enable:
                sum += int(match.group(1)) * int(match.group(2))
            continue

        if match.group(0) == "don't()":
            enable = False
            continue

        if match.group(0) == "do()":
            enable = True
            continue
        assert False
    print(f"sum is {sum}-----")
    return sum

def part2():

    sum = 0

    with open ("input.txt", 'r') as f:
        for line in f:
            sum += part2_line(line)
            
    print(sum)
#print(part2_line("don't()do(mul()mul(2,3)do()mul(2,3)mul(6,7)mul(don't()"))
part2()

File: 2024_d3/a.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <regex>
#include <stdint.h>

int main() {
  std::ifstream f("input.txt");
  std::string line;
  std::regex reg("mul\\(([0-9]{1,3}),([0-9]{1,3})\\)|do\\(\\)|don't\\(\\)");

  uint64_t res = 0;
  auto enable = true;
  while (std::getline(f, line)) {

    std::smatch sm;

    std::sregex_iterator it(line.begin(), line.end(), reg);
    for (auto end = std::sregex_iterator(); it != end; ++it) {

      if (it->str().find("mul") != std::string::npos) {

        if (enable) {
          int this_res = std::stoi((*it)[1]) * std::stoi((*it)[2]);
          res += this_res;
        }
        continue;
      }

      if (it->str() == "do()") {
        enable = true;
        continue;
      }
      if (it->str() == "don't()") {
        enable = false;
        continue;
      }

      assert(false);
    }
  }
  std::cout << res << std::endl;
}
